src/dbgshim/debugshim.cpp

```diff
@@ -291,39 +230,239 @@ STDMETHODIMP CLRDebuggingImpl::OpenVirtualProcess(
+//   pLibraryProvider - a callback for locating DBI and DAC
+//   dbiModulePath - returns the DBI module path
+//   dacModulePath - returns the DAC module path
+HRESULT CLRDebuggingImpl::ProvideLibraries(
+    ClrInfo& clrInfo,
+    ICLRDebuggingLibraryProvider3* pLibraryProvider,
+    SString& dbiModulePath,
+    SString& dacModulePath)
+{
+    HMODULE hDbi = NULL;
+    HMODULE hDac = NULL;
+    HRESULT hr = CLRDebuggingImpl::ProvideLibraries(clrInfo, pLibraryProvider, dbiModulePath, dacModulePath, &hDbi, &hDac);
+    if (SUCCEEDED(hr))
+    {
+        // The dbgshim create DBI instance APIs don't support just ICLRDebuggingLibraryProvider which is what
+        // it means if the handles returned are not null. At least ICLRDebuggingLibraryProvider2 is needed and
+        // ICLRDebuggingLibraryProvider3 for Unix platforms.
+        if (hDbi != NULL || hDac != NULL)
+        {
+            hr = E_INVALIDARG;
+        }
+    }
+    return hr;
+}
+
+// Call the library provider to get the DBI and DAC
+//
+// Arguments:
+//   clrInfo - the runtime info
+//   pLibraryProvider - a callback for locating DBI and DAC
+//   dbiModulePath - returns the DBI module path
+//   dacModulePath - returns the DAC module path
+//   phDbi - returns the DBI module handle if old library provider
+//   phDac - returns the DAC module handle if old library provider
+HRESULT CLRDebuggingImpl::ProvideLibraries(
+    ClrInfo& clrInfo,
+    IUnknown* punk,
+    SString& dbiModulePath,
+    SString& dacModulePath,
+    HMODULE* phDbi,
+    HMODULE* phDac)
+{
+    ReleaseHolder<ICLRDebuggingLibraryProvider3> pLibraryProvider3;
+    ReleaseHolder<ICLRDebuggingLibraryProvider2> pLibraryProvider2;
+    ReleaseHolder<ICLRDebuggingLibraryProvider> pLibraryProvider;
+    LPWSTR pDbiModulePath = NULL;
+    LPWSTR pDacModulePath = NULL;
+    HRESULT hr = S_OK;
+
+    _ASSERTE(punk != NULL);
+    _ASSERTE(phDbi != NULL);
+    _ASSERTE(phDac != NULL);
+
+    // Validate the incoming index info
+    if (!clrInfo.IsValid())
+    {
+        hr = CORDBG_E_INCOMPATIBLE_PROTOCOL;
+        goto exit;
+    }
+
+    if (SUCCEEDED(punk->QueryInterface(__uuidof(ICLRDebuggingLibraryProvider3), (void**)&pLibraryProvider3)))
+    {
+        const WCHAR* wszRuntimeModulePath = !clrInfo.RuntimeModulePath.IsEmpty() ? clrInfo.RuntimeModulePath.GetUnicode() : NULL;
+        if (clrInfo.WindowsTarget)
+        {
+            // Ask library provider for DBI
+            if (FAILED(pLibraryProvider3->ProvideWindowsLibrary(
+                clrInfo.DbiName, 
+                wszRuntimeModulePath,
+                clrInfo.IndexType,
+                clrInfo.DbiTimeStamp,
+                clrInfo.DbiSizeOfImage,
+                &pDbiModulePath)) || pDbiModulePath == NULL)
+            {
+                hr = CORDBG_E_LIBRARY_PROVIDER_ERROR;
+                goto exit;
+            }
+            // Ask library provider for DAC
+            if (FAILED(pLibraryProvider3->ProvideWindowsLibrary(
```

Comment:
We are not allowing for retarget the dac in provider 3?

Reply:
I'm wondering if we really need to do that anymore. I think all the remapping is for Framework builds though it isn't clear.